### PR TITLE
HPCC-21390 Remove default implementation of virtual functions from eclhelper.hpp

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
Tested all examples in regression suite and ecl/regress to ensure they still compile.  (And found 1 problem in the process).
@shamser please review

**Comment by richardkchapman** [comment]
Curious how you chose which to move (doesn't look like you moved all).

Do any of these changes cause existing compiled queries to need to be recompiled?

**Comment by ghalliday** [comment]
I don't think it will require any queries to be recompiled, but not 100% sure.  I think they should previously be implemented in the derived classes.  (With the hidden symbols enabled we didn't get link errors from queries.)

I tried to move them all.  I think the only ones remaining were virtual destructors and a couple in a class that wasn't an activity.  My main concern was to remove things like canMatch() which are potentially called for each row.  Any that are only called once per query will have little effect.

**Comment by ghalliday** [comment]
note, I didn't move inline functions - so that may be what you were seeing.

**Comment by richardkchapman** [comment]
It was ones in other include files (not eclhelper.hpp) where some had moved and some had not

File: common/thorhelper/thorcommon.ipp [hunk 1]
```diff
@@ -381,6 +384,10 @@ public:
     {
         return new CPrefixedRowDeserializer(offset, original->createInternalDeserializer(ctx, activityId));
     }
+    virtual void process(const byte * self, IFieldProcessor & target, unsigned from, unsigned to)
+    {
```
Comment (shamser) [inline common/thorhelper/thorcommon.ipp]: @ghalliday This definition for process()  seems to be different from the previous base class definition for process().  If this ticket simply removes default implementation of virtual functions, then I would have thought that this definition should reflect base class definition?

File: rtl/eclrtl/eclhelper_base.cpp [hunk 2]
```diff
@@ -526,9 +536,39 @@ const char * CThorCsvWriteArg::getCluster(unsigned idx) { return NULL; }
...
+
+unsigned CThorFetchArg::getFetchFlags() { return 0; }
+unsigned CThorFetchArg::getDiskFormatCrc() { return 0; }
+unsigned CThorFetchArg::getProjectedFormatCrc() { return getDiskFormatCrc(); }  // Should really be crc of queryProjectedDiskRecordSize layout
+void CThorFetchArg::getFileEncryptKey(size32_t & keyLen, void * & key) { keyLen = 0; key = 0; }
+unsigned __int64 CThorFetchArg::getRowLimit() { return (unsigned __int64) -1; }
+void CThorFetchArg::onLimitExceeded()         { }
+size32_t CThorFetchArg::extractJoinFields(ARowBuilder & rowBuilder, const void * _right) { return 0; }
+bool CThorFetchArg::extractAllJoinFields()    { return false; }
+IOutputMetaData * CThorFetchArg::queryExtractedSize() { return NULL; }
```
Comment (shamser) [inline rtl/eclrtl/eclhelper_base.cpp]: @ghalliday should nullptr be returned (instead of NULL) here and below

**Review by shamser — COMMENTED** [review]
@ghalliday

File: common/thorhelper/thorcommon.ipp (hunk 1 above)
Comment (ghalliday) [reply common/thorhelper/thorcommon.ipp]: You're right, this is a bug fix added at the same time.  I will move them to a different PR.

File: rtl/eclrtl/eclhelper_base.cpp (hunk 2 above)
Comment (ghalliday) [reply rtl/eclrtl/eclhelper_base.cpp]: Yes, but this was copied code, so fine to stay as NULL.

**Comment by ghalliday** [comment]
@shamser see replies

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.

**Comment by richardkchapman** [comment]
@ghalliday Did you determine whether this would cause any queries to need to be recompiled?

**Comment by ghalliday** [comment]
@richardkchapman I have verfied that if I build workunits with the previous code they still load in a system with the changes, and also none of the functions are exported from the dlls.  So I am fairly confident it will not cause queries to need to be recompiled.

**Pull request merged**